src/board.rs

```diff
@@ -0,0 +1,65 @@
+#[cfg(test)]
+mod test {}
+
+struct PlayerState {
+    king: u64,
+    queens: u64,
+    bishops: u64,
+    knights: u64,
+    rooks: u64,
+    pawns: u64,
+    king_castle: bool,
+    queen_castle: bool
+}
+
+impl PlayerState {
+    fn is_valid(&self) -> bool {
+        let mut valid = self.king.count_ones() == 1;
+        let add_board: u128 = self.king as u128 + self.queens as u128 + self.bishops as u128 + self.knights as u128 + self.rooks as u128 + self.pawns as u128;
+        valid &= add_board == self.all_pieces() as u128;
```

Comment:
i'm not 100% sure what's happening here? is this checking for two pieces being on the same spot? if so, it might be good to leave a note about that, and how/why it works

Reply:
that's exactly what it does, i'll add a doc comment explaining.